### PR TITLE
96/delegation

### PR DESCRIPTION
**Pull request by JustinMi**
**Update 11/12**
Not finished, but opening this pull request for now to add information/updates to it as I go
- after this is merged, people must re-migrate! I modified the answers migration

### CODE REVIEW

File: app/controllers/buildings_controller.rb [hunk 1]
```diff
@@ -8,4 +8,17 @@ def show
...
+
+  ##
+  # Given a building, and an email, create a new building_operator
+  # and mark the given question in the building as delegated, and return a new
+  # building instance.
+  #
+  def delegate_question(building, email)
+    building_operator = BuildingOperator.where(email: email).first_or_create
+    building.answers.where{ |a| a.predelegated? }.each do |answer|
+      answer = Answer.create!(building: building, question: answer.question, building_operator: building_operator)
```
Comment (HacksForHugs) [inline app/controllers/buildings_controller.rb]: so i don't think building_operator exists as a a column anymore.. Pull the latest changes from master and migrate. Arpan changed the association from answer to building_operator/asset_manager into a polymorphic association; I believe the columns are now `user_type` and `user_id`

File: app/models/answer.rb [hunk 2]
```diff
@@ -21,8 +21,35 @@ class Answer < ApplicationRecord
...
   end
+
+  ##
+  # When a user is in delegation mode and hits a checkbox next to a question, a
+  # request is sent to mark the question's answer as "predelegated", and the
+  # email of the delegated user is stored in the answer's text
+  #
+  def set_status_predelegated(email)
+    self.status = :predelegated
+    self.text = email
```
Comment (HacksForHugs) [inline app/models/answer.rb]: so these won't actually commit to the database I believe. I think you need to `.save`. Because of this, if any errors occur you actually need to propagate them back up so that the controller and send an error json message. This function may actually be better in the controller since you're directly making changes to the database off of a network request.

File: app/models/answer.rb [hunk 3]
```diff
@@ -21,8 +21,35 @@ class Answer < ApplicationRecord
...
+    self.status = :unanswered
+    self.text = ''
+  end
+
+  def set_status_answered
+    self.status = :answered
+  end
+
+  def set_status_delegated
+    self.status = :delegated
```
Comment (HacksForHugs) [inline app/models/answer.rb]: see above  comment: I don't think this code is actually committing changes.

File: config/routes.rb [hunk 4]
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
+  namespace :api do
+    get 'building_operator/create'
```
Comment (HacksForHugs) [inline config/routes.rb]: we don't want to expose this api, so can probs remove this. Building operators can only be created on delegation, which you handled in your code `BuildingOperator.where(email: email).first_or_create`. By adding a route for building operator creation, we essentially allow anyone who makes a json post request to create a new building operator which is kinda dangerous.

File: db/migrate/20171010181113_create_answers.rb [hunk 5]
```diff
@@ -1,7 +1,7 @@
 class CreateAnswers < ActiveRecord::Migration[5.1]
   def change
     create_table :answers do |t|
-      t.text :text
+      t.text :text, default: ''
```
Comment (HacksForHugs) [inline db/migrate/20171010181113_create_answers.rb]: LMAO ok i feel so bad for introducing this bad habit.... so you can make changes to these files UNLESS master has them. When rake db:migrate is run again, these files will not run again. Therefore, the only way to run these migration changes would be to drop the database and recreate, and reseed. This means we would have to do it on production too. You should never ever drop your database on production for migrations.

Comment (HacksForHugs) [reply db/migrate/20171010181113_create_answers.rb]: instead, create a new migration that changes the column info

File: app/controllers/buildings_controller.rb (hunk 1 above)
Comment (JustinMi) [reply app/controllers/buildings_controller.rb]: - [ ]

File: config/routes.rb (hunk 4 above)
Comment (JustinMi) [reply config/routes.rb]: Yah my b I forgot to remove this

File: app/models/answer.rb (hunk 2 above)
Comment (HacksForHugs) [reply app/models/answer.rb]: this should but be a put method in the controller I think, just because the input isn't controlled by us unlike set_status_delegated. `email` is something the user passes in, and it needs to be validated and potentially throw an error if the email passed in looks not like an email.

File: db/migrate/20171115045045_add_status_to_answers.rb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-class AddStatusEnumToAnswer < ActiveRecord::Migration[5.1]
+class AddStatusToAnswers < ActiveRecord::Migration[5.1]
```
Comment (HacksForHugs) [inline db/migrate/20171115045045_add_status_to_answers.rb]: why did you change this? this migration is already included in master so everyone already has it, which means it shouldn't be changed anymore

File: app/controllers/api/answers_controller.rb [hunk 7]
```diff
@@ -18,6 +18,15 @@ def update
     end
   end
 
+  ##
+  # Validates that the email is valid address, then sets the status of the
+  # answer to be predelegated
+  #
+  def predelegate_answer(email)
+    answer = Answer.find(params[:id])
+    answer.set_status_predelegated(email) if email =~ /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
```
Comment (HacksForHugs) [inline app/controllers/api/answers_controller.rb]: this might be better as a custom validation method in the answer model. So if status = pre delegated, text should be valid email.
validate answer.text, using custom validation, on update.

**Comment by HacksForHugs** [comment]
So basically this works for delegating questions from within a single building. Should an asset manager attempt to delegate questions across a category and multiple buildings, we will need a different function.

**Comment by JustinMi** [comment]
Should I implement that function? Or is it stretch feature
